Userland/Libraries/LibWeb/DOM/HTMLCollection.cpp

```diff
@@ -50,7 +50,7 @@ void HTMLCollection::visit_edges(Cell::Visitor& visitor)
         visitor.visit(element);
 }
 
-JS::MarkedVector<Element*> HTMLCollection::collect_matching_elements() const
+Vector<JS::NonnullGCPtr<Element>> const& HTMLCollection::collect_matching_elements() const
```

Comment:
A `Vector<JS::NonnullGCPtr>` doesn't protect its elements. Once we return it from here, it's no longer visible to the garbage collector.
The speedup is certainly attractive, but how can we know (at compile time) that this won't lead to a use-after-free?

Reply:
Good point. Tthe current users of this API just need to lookup a matching element. Perhaps I could provide a find_if style helper to pull a matching element out instead, and not expose this API publically?

Reply:
If that works, it sounds like a decent solution!

Reply:
HTMLFormControlsCollection is the niggly case, as it needs to find an element, but also check if there are more than one matching element. I'll think about finding a nice solution for that. Or just drop a FIXME for later on :D

Reply:
I've limited the optimization to within the class for now. I'm trying to find a nice method of exposing it outside of this class which doesn't have that pitfall, but I figure this is still an improvement as is